From 3cffb479b0d470863439be809c86d5b3687c6055 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Fri, 26 Apr 2024 17:15:29 +0200
Subject: [PATCH 618/841] feat(st-pmic): add support for STPMIC1L and STPMIC2L

Add the support for STPMIC1L and STPMIC2L.
- The BUCK1 can support an high voltage range like BUCK2 depending
on the NVM content.
- Those PMIC support GPO that are used to control external regulators.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git f428c3905924e4ff42a06b4dc864cd1dbb3b7e0f feat(st-pmic): add support for STPMIC1L and STPMIC2L]
Change-Id: I0ccd6aa0768873f8fc94fd514ddf70788ceea4bb
Signed-off-by: Pascal Paillet <p.paillet@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/378255
Reviewed-by: Pascal PAILLET <p.paillet@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Pascal PAILLET <p.paillet@st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 drivers/st/pmic/stm32mp_pmic2.c | 48 ++++++++++++++++++++++++++-
 drivers/st/pmic/stpmic2.c       | 57 ++++++++++++++++++++++++++-------
 include/drivers/st/stpmic2.h    | 45 ++++++++++++++++++++++++++
 3 files changed, 138 insertions(+), 12 deletions(-)

diff --git a/drivers/st/pmic/stm32mp_pmic2.c b/drivers/st/pmic/stm32mp_pmic2.c
index 9ab65ee13..27ac51206 100644
--- a/drivers/st/pmic/stm32mp_pmic2.c
+++ b/drivers/st/pmic/stm32mp_pmic2.c
@@ -39,6 +39,14 @@ static int dt_get_pmic_node(void *fdt)
 
 	if (node == -FDT_ERR_BADOFFSET) {
 		node = fdt_node_offset_by_compatible(fdt, -1, "st,stpmic2");
+		if (node < 0) {
+			node = fdt_node_offset_by_compatible(fdt, -1,
+							     "st,stpmic2l");
+		}
+		if (node < 0) {
+			node = fdt_node_offset_by_compatible(fdt, -1,
+							     "st,stpmic1l");
+		}
 	}
 
 	return node;
@@ -327,7 +335,7 @@ int stpmic2_set_prop(const struct regul_description *desc, uint16_t prop, uint32
 	return 0;
 }
 
-static struct regul_ops pmic2_ops = {
+static const struct regul_ops pmic2_ops = {
 	.set_state = pmic2_set_state,
 	.get_state = pmic2_get_state,
 	.set_voltage = pmic2_set_voltage,
@@ -336,6 +344,12 @@ static struct regul_ops pmic2_ops = {
 	.set_flag = pmic2_set_flag,
 };
 
+static const struct regul_ops pmic2_gpo_ops = {
+	.set_state = pmic2_set_state,
+	.get_state = pmic2_get_state,
+	.set_flag = pmic2_set_flag,
+};
+
 #define DEFINE_PMIC_REGUL_HANDLE(rid) \
 [(rid)] = { \
 	.id = (rid), \
@@ -343,6 +357,7 @@ static struct regul_ops pmic2_ops = {
 
 static struct regul_handle_s pmic2_regul_handles[STPMIC2_NB_REG] = {
 	DEFINE_PMIC_REGUL_HANDLE(STPMIC2_BUCK1),
+	DEFINE_PMIC_REGUL_HANDLE(STPMIC2_BUCK1H),
 	DEFINE_PMIC_REGUL_HANDLE(STPMIC2_BUCK2),
 	DEFINE_PMIC_REGUL_HANDLE(STPMIC2_BUCK3),
 	DEFINE_PMIC_REGUL_HANDLE(STPMIC2_BUCK4),
@@ -360,6 +375,12 @@ static struct regul_handle_s pmic2_regul_handles[STPMIC2_NB_REG] = {
 	DEFINE_PMIC_REGUL_HANDLE(STPMIC2_LDO8),
 
 	DEFINE_PMIC_REGUL_HANDLE(STPMIC2_REFDDR),
+
+	DEFINE_PMIC_REGUL_HANDLE(STPMIC2_GPO1),
+	DEFINE_PMIC_REGUL_HANDLE(STPMIC2_GPO2),
+	DEFINE_PMIC_REGUL_HANDLE(STPMIC2_GPO3),
+	DEFINE_PMIC_REGUL_HANDLE(STPMIC2_GPO4),
+	DEFINE_PMIC_REGUL_HANDLE(STPMIC2_GPO5),
 };
 
 #define DEFINE_REGUL(rid, name) \
@@ -369,8 +390,16 @@ static struct regul_handle_s pmic2_regul_handles[STPMIC2_NB_REG] = {
 	.driver_data = &pmic2_regul_handles[rid], \
 }
 
+#define DEFINE_GPO(rid, name) \
+[rid] = { \
+	.node_name = name, \
+	.ops = &pmic2_gpo_ops, \
+	.driver_data = &pmic2_regul_handles[rid], \
+}
+
 static const struct regul_description pmic2_descs[STPMIC2_NB_REG] = {
 	DEFINE_REGUL(STPMIC2_BUCK1, "buck1"),
+	DEFINE_REGUL(STPMIC2_BUCK1H, "buck1h"),
 	DEFINE_REGUL(STPMIC2_BUCK2, "buck2"),
 	DEFINE_REGUL(STPMIC2_BUCK3, "buck3"),
 	DEFINE_REGUL(STPMIC2_BUCK4, "buck4"),
@@ -388,6 +417,12 @@ static const struct regul_description pmic2_descs[STPMIC2_NB_REG] = {
 	DEFINE_REGUL(STPMIC2_LDO8, "ldo8"),
 
 	DEFINE_REGUL(STPMIC2_REFDDR, "refddr"),
+
+	DEFINE_GPO(STPMIC2_GPO1, "gpo1"),
+	DEFINE_GPO(STPMIC2_GPO2, "gpo2"),
+	DEFINE_GPO(STPMIC2_GPO3, "gpo3"),
+	DEFINE_GPO(STPMIC2_GPO4, "gpo4"),
+	DEFINE_GPO(STPMIC2_GPO5, "gpo5"),
 };
 
 static int register_pmic2(void)
@@ -418,6 +453,15 @@ static int register_pmic2(void)
 		int ret;
 		const fdt32_t *cuint;
 
+		if (strcmp(reg_name, "buck1") == 0) {
+			bool high;
+
+			ret = stpmic2_is_buck1_high_voltage(pmic2, &high);
+			if (high) {
+				reg_name = "buck1h";
+			}
+		}
+
 		for (i = 0; i < STPMIC2_NB_REG; i++) {
 			desc = &pmic2_descs[i];
 			if (strcmp(desc->node_name, reg_name) == 0) {
@@ -487,6 +531,8 @@ void initialize_pmic(void)
 	}
 	INFO("PMIC2 product ID = 0x%02x\n", val);
 
+	pmic2->ref_id = ((val & PMIC_REF_ID_MASK) >> PMIC_REF_ID_SHIFT);
+
 	ret = register_pmic2();
 	if (ret < 0) {
 		ERROR("Register pmic2 failed\n");
diff --git a/drivers/st/pmic/stpmic2.c b/drivers/st/pmic/stpmic2.c
index 075f62dcf..cd0eca97a 100644
--- a/drivers/st/pmic/stpmic2.c
+++ b/drivers/st/pmic/stpmic2.c
@@ -37,7 +37,7 @@ struct regul_struct {
 };
 
 /* Voltage tables in mV */
-static const uint16_t buck1236_volt_table[] = {
+static const uint16_t buck_low_volt_table[] = {
 	500U, 510U, 520U, 530U, 540U, 550U, 560U, 570U, 580U, 590U,
 	600U, 610U, 620U, 630U, 640U, 650U, 660U, 670U, 680U, 690U,
 	700U, 710U, 720U, 730U, 740U, 750U, 760U, 770U, 780U, 790U,
@@ -53,7 +53,7 @@ static const uint16_t buck1236_volt_table[] = {
 	1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U
 };
 
-static const uint16_t buck457_volt_table[] = {
+static const uint16_t buck_high_volt_table[] = {
 	1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U,
 	1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U,
 	1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U, 1500U,
@@ -119,6 +119,14 @@ static const uint16_t refddr_volt_table[] = {
 	.ocp_mask		= FS_OCP_ ## ID, \
 }
 
+#define DEFINE_GPOx(regu_name, ID) { \
+	.name			= regu_name, \
+	.en_cr			= ID ## _MAIN_CR, \
+	.alt_en_cr		= ID ## _ALT_CR, \
+	.msrt_reg		= GPO_MRST_CR, \
+	.msrt_mask		= ID ## _MRST, \
+}
+
 #define DEFINE_REFDDR(regu_name, ID, table) { \
 	.name			= regu_name, \
 	.volt_table		= table, \
@@ -137,19 +145,21 @@ static const uint16_t refddr_volt_table[] = {
 /* Table of Regulators in PMIC SoC */
 static const struct regul_struct regul_table[STPMIC2_NB_REG] = {
 	[STPMIC2_BUCK1] = DEFINE_BUCK("buck1", BUCK1, BUCKS_PD_CR1,
-				      buck1236_volt_table),
+				      buck_low_volt_table),
+	[STPMIC2_BUCK1H] = DEFINE_BUCK("buck1h", BUCK1, BUCKS_PD_CR1,
+				      buck_high_volt_table),
 	[STPMIC2_BUCK2] = DEFINE_BUCK("buck2", BUCK2, BUCKS_PD_CR1,
-				      buck1236_volt_table),
+				      buck_low_volt_table),
 	[STPMIC2_BUCK3] = DEFINE_BUCK("buck3", BUCK3, BUCKS_PD_CR1,
-				      buck1236_volt_table),
+				      buck_low_volt_table),
 	[STPMIC2_BUCK4] = DEFINE_BUCK("buck4", BUCK4, BUCKS_PD_CR1,
-				      buck457_volt_table),
+				      buck_high_volt_table),
 	[STPMIC2_BUCK5] = DEFINE_BUCK("buck5", BUCK5, BUCKS_PD_CR2,
-				      buck457_volt_table),
+				      buck_high_volt_table),
 	[STPMIC2_BUCK6] = DEFINE_BUCK("buck6", BUCK6, BUCKS_PD_CR2,
-				      buck1236_volt_table),
+				      buck_low_volt_table),
 	[STPMIC2_BUCK7] = DEFINE_BUCK("buck7", BUCK7, BUCKS_PD_CR2,
-				      buck457_volt_table),
+				      buck_high_volt_table),
 
 	[STPMIC2_REFDDR] = DEFINE_REFDDR("refddr", REFDDR, refddr_volt_table),
 
@@ -162,6 +172,12 @@ static const struct regul_struct regul_table[STPMIC2_NB_REG] = {
 	[STPMIC2_LDO7] = DEFINE_LDOx("ldo7", LDO7, ldo235678_volt_table),
 	[STPMIC2_LDO8] = DEFINE_LDOx("ldo8", LDO8, ldo235678_volt_table),
 
+	[STPMIC2_GPO1] = DEFINE_GPOx("gpo1", GPO1),
+	[STPMIC2_GPO2] = DEFINE_GPOx("gpo2", GPO2),
+	[STPMIC2_GPO3] = DEFINE_GPOx("gpo3", GPO3),
+	[STPMIC2_GPO4] = DEFINE_GPOx("gpo4", GPO4),
+	[STPMIC2_GPO5] = DEFINE_GPOx("gpo5", GPO5),
+
 };
 
 int stpmic2_register_read(struct pmic_handle_s *pmic,
@@ -356,7 +372,8 @@ int stpmic2_regulator_get_prop(struct pmic_handle_s *pmic, uint8_t id,
 	switch (prop) {
 	case STPMIC2_BYPASS:
 		if ((id <= STPMIC2_BUCK7) || (id == STPMIC2_LDO1) ||
-		    (id == STPMIC2_LDO4) || (id == STPMIC2_REFDDR)) {
+		    (id == STPMIC2_LDO4) || (id == STPMIC2_REFDDR) ||
+		    (pmic->ref_id != PMIC_REF_ID_STPMIC25)) {
 			return 0;
 		}
 
@@ -399,7 +416,8 @@ int stpmic2_regulator_set_prop(struct pmic_handle_s *pmic, uint8_t id,
 					       regul->msrt_mask);
 	case STPMIC2_BYPASS:
 		if ((id <= STPMIC2_BUCK7) || (id == STPMIC2_LDO1) ||
-		    (id == STPMIC2_LDO4) || (id == STPMIC2_REFDDR)) {
+		    (id == STPMIC2_LDO4) || (id == STPMIC2_REFDDR) ||
+		    (pmic->ref_id != PMIC_REF_ID_STPMIC25)) {
 			return RET_ERROR_NOT_SUPPORTED;
 		}
 
@@ -472,3 +490,20 @@ int stpmic2_get_product_id(struct pmic_handle_s *pmic, uint8_t *val)
 {
 	return stpmic2_register_read(pmic, PRODUCT_ID, val);
 }
+
+int stpmic2_is_buck1_high_voltage(struct pmic_handle_s *pmic, bool *high)
+{
+	int ret;
+	uint8_t val = 0;
+
+	if (pmic->ref_id == PMIC_REF_ID_STPMIC25) {
+		*high = false;
+		return 0;
+	}
+
+	ret = stpmic2_register_read(pmic, NVM_BUCK1_VOUT_SHR, &val);
+
+	*high = val & BUCK1_VRAN_GE_CFG;
+
+	return ret;
+}
diff --git a/include/drivers/st/stpmic2.h b/include/drivers/st/stpmic2.h
index bd04c4ec2..c84f7703a 100644
--- a/include/drivers/st/stpmic2.h
+++ b/include/drivers/st/stpmic2.h
@@ -12,6 +12,7 @@
 
 enum {
 	STPMIC2_BUCK1 = 0,
+	STPMIC2_BUCK1H,
 	STPMIC2_BUCK2,
 	STPMIC2_BUCK3,
 	STPMIC2_BUCK4,
@@ -27,6 +28,11 @@ enum {
 	STPMIC2_LDO6,
 	STPMIC2_LDO7,
 	STPMIC2_LDO8,
+	STPMIC2_GPO1,
+	STPMIC2_GPO2,
+	STPMIC2_GPO3,
+	STPMIC2_GPO4,
+	STPMIC2_GPO5,
 	STPMIC2_NB_REG
 };
 
@@ -58,6 +64,7 @@ enum {
 #define BUCKS_PD_CR2		0x1A
 #define LDOS_PD_CR1		0x1B
 #define LDOS_PD_CR2		0x1C
+#define GPO_MRST_CR		0x1C
 #define BUCKS_MRST_CR		0x1D
 #define LDOS_MRST_CR		0x1E
 /* Buck CR */
@@ -96,6 +103,13 @@ enum {
 #define BUCK7_ALT_CR1		0x40
 #define BUCK7_ALT_CR2		0x41
 #define BUCK7_PWRCTRL_CR	0x42
+/* GPO CR used only on PMIC1L and PMIC2L */
+#define GPO1_MAIN_CR		0x43
+#define GPO1_ALT_CR		0x44
+#define GPO1_PWRCTRL_CR		0x45
+#define GPO2_MAIN_CR		0x46
+#define GPO2_ALT_CR		0x47
+#define GPO2_PWRCTRL_CR		0x48
 /* LDO CR */
 #define LDO1_MAIN_CR		0x4C
 #define LDO1_ALT_CR		0x4D
@@ -124,6 +138,16 @@ enum {
 #define REFDDR_MAIN_CR		0x64
 #define REFDDR_ALT_CR		0x65
 #define REFDDR_PWRCTRL_CR	0x66
+/* GPO CR used only on PMIC1L and PMIC2L */
+#define GPO3_MAIN_CR		0x67
+#define GPO3_ALT_CR		0x68
+#define GPO3_PWRCTRL_CR		0x69
+#define GPO4_MAIN_CR		0x6A
+#define GPO4_ALT_CR		0x6B
+#define GPO4_PWRCTRL_CR		0x6C
+#define GPO5_MAIN_CR		0x6D
+#define GPO5_ALT_CR		0x6E
+#define GPO5_PWRCTRL_CR		0x6F
 /* INTERRUPT CR */
 #define INT_PENDING_R1		0x70
 #define INT_PENDING_R2		0x71
@@ -146,9 +170,18 @@ enum {
 #define INT_DBG_LATCH_R3	0x82
 #define INT_DBG_LATCH_R4	0x83
 /* NVM SHADOW REGISTERS */
+#define NVM_BUCK1_VOUT_SHR	0x9C
 #define NVM_LDOS_IOUT_SHR	0xAE
 #define LDO5_ILIM_SHIFT		2
 
+/* PRODUCT_ID bits definition */
+#define PMIC_NVM_ID_MASK	GENMASK_32(3, 0)
+#define PMIC_REF_ID_MASK	GENMASK_32(7, 4)
+#define PMIC_REF_ID_SHIFT	4
+#define PMIC_REF_ID_STPMIC1L	U(1)
+#define PMIC_REF_ID_STPMIC25	U(2)
+#define PMIC_REF_ID_STPMIC2L	U(3)
+
 /* BUCKS_MRST_CR bits definition */
 #define BUCK1_MRST		BIT(0)
 #define BUCK2_MRST		BIT(1)
@@ -169,6 +202,13 @@ enum {
 #define LDO7_MRST		BIT(6)
 #define LDO8_MRST		BIT(7)
 
+/* GPO_MRST_CR bits definition */
+#define GPO1_MRST		BIT(1)
+#define GPO2_MRST		BIT(2)
+#define GPO3_MRST		BIT(3)
+#define GPO4_MRST		BIT(4)
+#define GPO5_MRST		BIT(5)
+
 /* LDOx_MAIN_CR */
 #define LDO_VOLT_SHIFT		1
 #define LDO_BYPASS		BIT(6)
@@ -240,6 +280,9 @@ enum {
 #define FS_OCP_LDO7		BIT(6)
 #define FS_OCP_LDO8		BIT(7)
 
+/* NVM_BUCK1_VOUT_SHR */
+#define BUCK1_VRAN_GE_CFG	BIT(7)
+
 /* IRQ definitions */
 #define IT_PONKEY_F	0
 #define IT_PONKEY_R	1
@@ -272,6 +315,7 @@ struct pmic_handle_s {
 	struct i2c_handle_s *i2c_handle;
 	uint32_t i2c_addr;
 	unsigned int pmic_status;
+	uint8_t ref_id;
 };
 
 int stpmic2_register_read(struct pmic_handle_s *pmic,
@@ -307,4 +351,5 @@ int stpmic2_regulator_get_prop(struct pmic_handle_s *pmic, uint8_t id,
 int stpmic2_regulator_set_prop(struct pmic_handle_s *pmic, uint8_t id,
 			       enum stpmic2_prop_id prop, uint32_t arg);
 
+int stpmic2_is_buck1_high_voltage(struct pmic_handle_s *pmic, bool *high);
 #endif /*STPMIC2_H*/
-- 
2.47.2

